File: zadanie4/zadanie4.cpp
#include <iostream>
#include <vector>
#include <cstdint>

using namespace std;

void input(int &n, int &m, vector<int> &chisla);
pair<int, int64_t> selectBestMove(int pos, int m, int lastMove, const vector<int> &chisla);
void playGame(const vector<int> &chisla, int m, int64_t &pavel, int64_t &vika);

int main() {
    int n{};
    int m{};
    int64_t pavel{};
    int64_t vika{};
    vector<int> chisla;
    input(n, m, chisla);
    playGame(chisla, m, pavel, vika);
    cout << (pavel > vika ? 1 : 0) << endl;
    return 0;
}

void input(int &n, int &m, vector<int> &chisla) {
    cin >> n >> m;
    chisla.resize(n);
    for (int i = 0; i < n; ++i){
        cin >> chisla[i];
    }
}

pair<int, int64_t> selectBestMove(int pos, int m, int lastMove, const vector<int> &chisla) {
    int bestK = 1;
    int64_t bestSum = INT64_MIN;
    int64_t currentSum = 0;
    for (int k = 1; k <= m && pos + k <= (int)chisla.size(); ++k) {
        if (k == lastMove){
            continue;
        } 
        currentSum += chisla[pos + k - 1];
        if (currentSum > bestSum) {
            bestSum = currentSum;
            bestK = k;
        }
    }
    return {bestK, bestSum};
}

void playGame(const vector<int> &chisla, int m, int64_t &pavel, int64_t &vika) {
    int pos = 0;
    int lastMove = -1;
    bool isPavel = true;
    pavel = vika = 0;
    while (pos < (int)chisla.size()) {
        pair<int, int64_t> moveResult = selectBestMove(pos, m, lastMove, chisla);
        int bestK = moveResult.first;
        int64_t bestSum = moveResult.second;
        if (isPavel){
            pavel += bestSum;
        } 
        else{
            vika += bestSum;
        } 
        pos += bestK;
        lastMove = bestK;
        isPavel = !isPavel;
    }
}

File: zadanie4/go4/zadanie4.go
package main

import "fmt"

func main() {
	_, m, chisla := input()
	pavel, vika := playGame(chisla, m)
	if pavel > vika {
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
}

func input() (n int, m int, chisla []int) {
	fmt.Scan(&n, &m)
	chisla = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&chisla[i])
	}
	return
}

func selectBestMove(pos, m, lastMove int, chisla []int) (bestK int, bestSum int64) {
	bestK = 1
	bestSum = -1 << 62
	currentSum := int64(0)
	for k := 1; k <= m && pos+k <= len(chisla); k++ {
		if k == lastMove {
			continue
		}
		currentSum += int64(chisla[pos+k-1])
		if currentSum > bestSum {
			bestSum = currentSum
			bestK = k
		}
	}
	return
}

func playGame(chisla []int, m int) (pavel, vika int64) {
	pos := 0
	lastMove := -1
	isPavel := true
	pavel, vika = 0, 0
	for pos < len(chisla) {
		bestK, bestSum := selectBestMove(pos, m, lastMove, chisla)
		if isPavel {
			pavel += bestSum
		} else {
			vika += bestSum
		}
		pos += bestK
		lastMove = bestK
		isPavel = !isPavel
	}
	return
}
